src/bsoncxx/test/version.cpp

```diff
@@ -0,0 +1,100 @@
+// Copyright 2009-present MongoDB, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <bsoncxx/private/version.hh>
+
+//
+
+#include <string>
+#include <vector>
+
+#include <bsoncxx/stdx/string_view.hpp>
+
+#include <bsoncxx/private/bson.hh>
+#include <bsoncxx/private/config/config.hh>
+
+#include <bsoncxx/test/catch.hh>
+
+namespace {
+
+using bsoncxx::test_util::split_version;
+
+TEST_CASE("split_version", "[bsoncxx][test][split_version]") {
+    CHECK(split_version("1") == std::vector<int>({1}));
+    CHECK(split_version("1.2") == std::vector<int>({1, 2}));
+    CHECK(split_version("1.2.3") == std::vector<int>({1, 2, 3}));
+    CHECK(split_version("11.22.33") == std::vector<int>({11, 22, 33}));
+
+    CHECK(split_version("1-") == std::vector<int>({1}));
+    CHECK(split_version("1.2-") == std::vector<int>({1, 2}));
+    CHECK(split_version("1.2.3-") == std::vector<int>({1, 2, 3}));
+    CHECK(split_version("1.2.3-suffix") == std::vector<int>({1, 2, 3}));
+}
+
+TEST_CASE("bson version numbers", "[bsoncxx][test]") {
+    auto expect_str = bsoncxx::stdx::string_view{BSON_REQUIRED_VERSION()};
+
+    CAPTURE(expect_str);
+
+    auto const expect = split_version(expect_str);
+
+    // The major version number should always be present.
+    REQUIRE(expect.size() > 0);
+
+    CHECK(bson_get_major_version() == expect[0]);
+
+    CHECKED_IF(expect.size() > 1) {
+        CHECK(bson_get_minor_version() >= expect[1]);
+    }
+
+    CHECKED_IF(expect.size() > 2) {
+        CHECK(bson_get_micro_version() >= expect[2]);
```

Comment:
To handle a case like `expect=1.2.3 actual=1.3.2`:
```suggestion
        if (bson_get_minor_version() == expect[1]) {
            // For same major.minor, check patch.
            CHECK(bson_get_micro_version() >= expect[2]);
        }
```

Similar suggestion applies for the other patch version checks.

Reply:
This was deliberate to support running tests against C Driver libraries regardless whether they are imported via `find_package()` or via `add_subdirectory()`, both of which may  refer to a version than the C++ Driver being tested, such as in our EVG tasks (e.g. `-dev`). Because `1.3.2` is compatible with a requirement of `1.2.3`, it is also permitted by these tests. (That is, if `find_package(bson 1.2.3)` [accepts](https://github.com/mongodb/mongo-cxx-driver/blob/f1c276f7e26b29063d2ec31bc85d6627fd07d38a/src/bsoncxx/cmake/CMakeLists.txt#L22) `1.3.2`, so should these tests.)

Reply:
Whoops, I misinterpreted the suggestion as modifying the minor version number comparison, but you meant guarding the patch version number comparison. My mistake.